yarn-project/txe/src/util/simulation_error.ts

```diff
@@ -0,0 +1,79 @@
+import { type SimulationError, isNoirCallStackUnresolved } from '@aztec/circuit-types';
+import { AztecAddress, Fr, FunctionSelector, PUBLIC_DISPATCH_SELECTOR } from '@aztec/circuits.js';
+import { type Logger } from '@aztec/foundation/log';
+import { type ContractDataOracle } from '@aztec/pxe';
+import { resolveAssertionMessage, resolveOpcodeLocations } from '@aztec/simulator';
+
+import { type TXEDatabase } from './txe_database.js';
+
+export async function enrichPublicSimulationError(
+  err: SimulationError,
+  db: TXEDatabase,
+  contractDataOracle: ContractDataOracle,
+  logger: Logger,
+) {
+  // Try to fill in the noir call stack since the PXE may have access to the debug metadata
+  const callStack = err.getCallStack();
+  const originalFailingFunction = callStack[callStack.length - 1];
+  // TODO(https://github.com/AztecProtocol/aztec-packages/issues/8985): Properly fix this.
+  // To be able to resolve the assertion message, we need to use the information from the public dispatch function,
+  // no matter what the call stack selector points to (since we've modified it to point to the target function).
+  // We should remove this because the AVM (or public protocol) shouldn't be aware of the public dispatch calling convention.
+  const debugInfo = await contractDataOracle.getFunctionDebugMetadata(
```

Comment:
Would it be possible to have this version and the non-txe version of error enriching unified? Since we want to rework it soon (to have rethrowing) it's probably better if it's not duplicated?

Reply:
Done! I don't really like the way I split it, but since we're gonna be changhing this anyway...